File: cracking_the_coding_interview/ctci-contacts.py
class Node:

    def __init__(self):
        self.letters = {}
        self.uses = 0

class ContactList:

    def __init__(self):
        self.root = Node()

    """
        Insert a new contact name in the tree of contacts
    """
    def insert_contact(self, name, l = None, i = 0):
        if not l:
            l = self.root

        # if name is empty, quit
        if i >= len(name):
            return

        # insert letter by letter.
        # if the next letter is not in the letters of this node,
        # add the letter
        if name[i] not in l.letters:
            # if this is the last letter
            l.letters[name[i]] = Node()
        # increment the uses of this node
        l.letters[name[i]].uses += 1

        # call recursively this method to insert fully the contact name
        # with the rest of the name
        return self.insert_contact(name, l.letters[name[i]], i+1)


    """
        Down the tree following the partial string
        and return the number of uses of this node
    """
    def count_contact(self, partial):
        if not partial:
            return

        node = self.root
        i = 0
        while i < len(partial):
            if partial[i] not in node.letters:
                # the partial match 0 contact
                return 0

            node = node.letters[partial[i]]
            i += 1

        return node.uses


# create the contact list
contact_list = ContactList()

# read each operations
n = int(input().strip())
for _ in range(n):
    op, contact = input().strip().split(' ')

    if op == "add":
        contact_list.insert_contact(contact)
    else:
        print(contact_list.count_contact(contact))

File: cracking_the_coding_interview/ctci-contacts.java
import java.util.Scanner;
import java.util.HashMap;

public class Solution {

    private static class Node {

        private int uses;
        private HashMap<Character, Node> children = new HashMap<>();

        public Node() {
            this.uses = 0;
        }

        public int getUses() {
            return this.uses;
        }

        public HashMap getChildren() {
            return this.children;
        }

        public Node getChild(Character ch) {
            return this.children.get(ch);
        }

        public void addUse() {
            this.uses++;
        }

    }

    private static class ContactTree {

        private Node root;

        public ContactTree() {
            this.root = new Node();
        }

        /**
         * Insert a new contact name in the contact tree
         *
         * @param String name : the contact name
         * @return void
         */
        public void insertContact(String name) {
            Node current = this.root;
            // put each Character of the name in the tree
            for (int i = 0; i < name.length(); i++) {
                Character ch = name.charAt(i);
                // if this Character is not a child of this Node,
                // add it
                current.getChildren().putIfAbsent(ch, new Node());
                // down the tree
                current = current.getChild(ch);
                // increment the number of uses of the current Node
                current.addUse();
            }
        }

        /**
         * Down the tree following the partial string
         * and return the number of uses of this node
         *
         * @param String partial
         * @return int : the number of name starting by partial
         */
        public int countContacts(String partial) {
            Node current = this.root;
            // go down the tree
            for (int i = 0; i < partial.length(); i++) {
                Character ch = partial.charAt(i);
                // if this Character is not a child of the current node,
                // return 0 (the partial matches 0 name)
                current = current.getChild(ch);
                if (current == null) {
                    return 0;
                }
            }
            
            return current.getUses();
        }

    }

    public static void main(String[] args) {
        ContactTree contactTree = new ContactTree();

        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        for (int i = 0; i < n; i++){
            String op = in.next();
            String contact = in.next();

            if (op.equals("add")) {
                contactTree.insertContact(contact);
            } else {
                System.out.println(contactTree.countContacts(contact));
            }

        }
    }
}
